pkg/dockerregistry/server/writelimiter/semaphore.go

```diff
@@ -0,0 +1,24 @@
+package writelimiter
+
+type semaphore struct {
+	c chan struct{}
+}
+
+func newSemaphore(count int) *semaphore {
+	return &semaphore{
+		c: make(chan struct{}, count),
+	}
+}
+
+func (s *semaphore) Up() {
+	<-s.c
+}
+
+func (s *semaphore) TryDown() bool {
```

Comment:
I see that you need to return a cancel func.  Context already does much of this, so we should be using a context.WithDeadline() and then passing that down.  The `try` func is not necessary.